test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BufferAdapter.scala

```diff
@@ -6,8 +6,15 @@ sealed abstract class BufferAdapter[BT <: Buffer, ET] {
   type BufferType = BT
   type ElementType = ET
 
-  def slice(): BufferType
-  def duplicate(): BufferType
+  /* Some methods have a Chain suffix because they are declared as abstract in
+   * java.nio.Buffer since Java 9, but with a result type of `Buffer` instead
+   * of the more specific `BufferType`. We use the `Chain` variant to be able
+   * to chain their application with further operations on the specific
+   * `BufferType`.
+   */
+
+  def sliceChain(): BufferType
+  def duplicateChain(): BufferType
```

Comment:
I don't get this. `BufferAdapter` does not have any inheritance relation to `Buffer`.

Reply:
The methods of `BufferAdapter` are pimped on the abstract type `BufferType`, which itself is `<: Buffer`. So you have code like `myBuffer.slice()` with `myBuffer: BufferType` which used to resolve to `bufferTypeToAdapter(myBuffer).slice()` (with result type `BufferType`) but now they don't go through the implicit conversion anymore, which means their result type is just `Buffer`.

Reply:
Ah, I see :P